File: go/topo_sort/topo_sort_test.go
package topo_sort

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func TestExample(t *testing.T) {
	assert.Equal(t, []int{3, 0, 2, 1}, TopoSort([][]int{{2, 1}, {}, {1}, {0, 2}}))

	assert.Equal(t, 0, len(TopoSort([][]int{{2, 1}, {3}, {1}, {0, 2}})))
	assert.Equal(t, 0, len(TopoSort([][]int{{0}})))
	assert.Equal(t, 0, len(TopoSort([][]int{{1}, {0}})))
}

func TestRandomCases(t *testing.T) {
	for range 10_000 {
		td := newRandomTestData()

		actual := TopoSort(td.adj)

		if td.hasCycle {
			assert.Equal(t, 0, len(actual))
			continue
		}

		n := len(td.graph)
		assert.Equal(t, n, len(actual))

		order := make([]int, n)
		for i := range n {
			order[actual[i]] = i
		}

		for u := range n {
			for _, v := range td.graph[u] {
				if !(order[u] < order[v]) {
					assert.FailNowf(t, "",
						"TopoSort(%v):\n  Node %v should come before node %v in the result!\n  Result: %v",
						td.graph, u, v, actual,
					)
				}
			}
		}
	}
}

File: python/topo_sort/topo_sort_test.py
from topo_sort import topo_sort
from topo_sort_test_data import RandomTestData


def test_example():
	assert topo_sort([[2, 1], [], [1], [0, 2]]) == [3, 0, 2, 1]

	assert len(topo_sort([[2, 1], [3], [1], [0, 2]])) == 0
	assert len(topo_sort([[0]])) == 0
	assert len(topo_sort([[1], [0]])) == 0


def test_random_cases():
	for _ in range(10_000):
		td = RandomTestData()

		actual = topo_sort(td.adj)

		if td.hasCycle:
			assert len(actual) == 0
			continue

		n = len(td.graph)
		assert len(actual) == n

		order = [0] * n
		for i in range(n):
			order[actual[i]] = i

		for u in range(n):
			for v in td.graph[u]:
				assert order[u] < order[v], \
					f'topo_sort({td.graph}):\n  Node {u} should come before node {v} in the result!\n  Result: {actual}'
